scylla-cql/src/frame/response/result.rs

```diff
@@ -856,6 +1052,139 @@ fn deser_result_metadata(
+                    // 1) no cached_metadata provided. Server is supposed to provide the result metadata.
+                    // 2) cached metadata present (so we should have asked for skipping metadata),
+                    //    but the server sent result metadata anyway.
+                    // In case 1 we have to deserialize result metadata. In case 2 we choose to do that,
+                    // too, because it's suspicious, so we had better use the new metadata just in case.
+                    // Also, we simply need to advance the buffer pointer past metadata, and this requires
+                    // parsing metadata.
+                    let server_metadata = {
+                        let global_table_spec = self
+                            .global_tables_spec
+                            .then(|| deser_table_spec(frame_slice.as_slice_mut()))
+                            .transpose()
+                            .map_err(ResultMetadataParseError::from)?;
+
+                        let col_specs = $deser_col_specs(
+                            frame_slice.as_slice_mut(),
+                            global_table_spec,
+                            self.col_count,
+                        )
+                        .map_err(ResultMetadataParseError::from)?;
+
+                        ResultMetadata {
+                            col_count: self.col_count,
+                            col_specs,
+                        }
+                    };
+                    if server_metadata.col_count() != server_metadata.col_specs().len() {
+                        return Err(RowsParseError::ColumnCountMismatch {
+                            col_count: server_metadata.col_count(),
+                            col_specs_count: server_metadata.col_specs().len(),
+                        });
+                    }
+                    ResultMetadataHolder::BorrowedOrOwned(Cow::Owned(server_metadata))
+                }
+            };
+
+            let rows_count: usize = types::read_int_length(frame_slice.as_slice_mut())
+                .map_err(RowsParseError::RowsCountParseError)?;
+
+            Ok(DeserializedMetadataAndRawRows {
+                metadata,
+                rows_count,
+                raw_rows: $raw_rows_constructor(frame_slice),
+            })
+        }
+    };
+}
+
+impl RawMetadataAndRawRows {
+    generate_deserialize_metadata!(
+        deserialize_borrowed_metadata,
+        '_,
+        RawRowsBorrowed<'_>,
+        RawRowsBorrowed,
+        |cached| ResultMetadataHolder::BorrowedOrOwned(Cow::Borrowed(cached)),
+        deser_col_specs_borrowed,
+    );
+
+    generate_deserialize_metadata!(
+        deserialize_owned_metadata,
+        'static,
+        RawRowsOwned,
+        |frame_slice: FrameSlice| RawRowsOwned(frame_slice.to_bytes()),
+        |cached| ResultMetadataHolder::SharedCached(Arc::clone(cached)),
+        deser_col_specs_owned,
+    );
+}
+
 fn deser_prepared_metadata(
```

Comment:
> Imo 2 repetitions don't justify introducing a macro and all of problems related to it.

When the repeated code is big enough (and this is the case), then I strongly believe it is justified.

In this case I can't deduplicate parts using normal functions because of different ownership/lifetimes. I encourage you to try yourself; I still believe the present state is the best approach here.

Reply:
Any future work on this code will require editing a macro, which is a pain because of much worse IDE (rust-analyzer) support compared to normal code, and much worse readability.

Reply:
Any future work on repeated code will require editing two places, which is a pain because it's extremely easy to forgot about code flow used by another use case.
Moreover, function forcibly deduplicated by extracting some parts tend to be much less readable than those which aren't that much modularised.

Reply:
It is possible to avoid macros here:
```rust
/// Deserializes ResultMetadata in the form mentioned by its name,
/// and deserializes rows count. Keeps rows in the serialized form.
///
/// If metadata is cached (in the PreparedStatement), it is reused (shared) from cache
/// instead of deserializing.
pub fn deserialize_metadata_generic<'this, 'result, T: RawRowsKind>(
    this: &'this RawMetadataAndRawRows,
    raw_rows_constructor: impl Fn(FrameSlice<'this>) -> T,
    use_cached_metadata: impl Fn(&'this Arc<ResultMetadata<'static>>) -> ResultMetadataHolder<'result>,
    deser_col_specs: impl Fn(
        &mut &'this [u8],
        Option<TableSpec<'this>>,
        usize,
    ) -> StdResult<Vec<ColumnSpec<'result>>, ColumnSpecParseError>,
) -> StdResult<DeserializedMetadataAndRawRows<'result, T>, RowsParseError> {
    let mut frame_slice = FrameSlice::new(&this.raw_metadata_and_rows);

    let metadata = match this.cached_metadata.as_ref() {
        Some(cached) if this.no_metadata => {
            // Server sent no metadata, but we have metadata cached. This means that we asked the server
            // not to send metadata in the response as an optimization. We use cached metadata instead.
            //$use_cached_metadata(cached)
            use_cached_metadata(cached)
        }
        None if this.no_metadata => {
            // Server sent no metadata and we have no metadata cached. Having no metadata cached,
            // we wouldn't have asked the server for skipping metadata. Therefore, this is most probably
            // not a SELECT, because in such case the server would send empty metadata both in Prepared
            // and in Result responses.
            ResultMetadataHolder::BorrowedOrOwned(Cow::Owned(ResultMetadata::mock_empty()))
        }
        Some(_) | None => {
            // Two possibilities:
            // 1) no cached_metadata provided. Server is supposed to provide the result metadata.
            // 2) cached metadata present (so we should have asked for skipping metadata),
            //    but the server sent result metadata anyway.
            // In case 1 we have to deserialize result metadata. In case 2 we choose to do that,
            // too, because it's suspicious, so we had better use the new metadata just in case.
            // Also, we simply need to advance the buffer pointer past metadata, and this requires
            // parsing metadata.
            let server_metadata = {
                let global_table_spec = this
                    .global_tables_spec
                    .then(|| deser_table_spec(frame_slice.as_slice_mut()))
                    .transpose()
                    .map_err(ResultMetadataParseError::from)?;

                let col_specs = deser_col_specs(
                    frame_slice.as_slice_mut(),
                    global_table_spec,
                    this.col_count,
                )
                .map_err(ResultMetadataParseError::from)?;

                ResultMetadata {
                    col_count: this.col_count,
                    col_specs,
                }
            };
            if server_metadata.col_count() != server_metadata.col_specs().len() {
                return Err(RowsParseError::ColumnCountMismatch {
                    col_count: server_metadata.col_count(),
                    col_specs_count: server_metadata.col_specs().len(),
                });
            }
            ResultMetadataHolder::BorrowedOrOwned(Cow::Owned(server_metadata))
        }
    };

    let rows_count: usize = types::read_int_length(frame_slice.as_slice_mut())
        .map_err(RowsParseError::RowsCountParseError)?;

    Ok(DeserializedMetadataAndRawRows {
        metadata,
        rows_count,
        raw_rows: raw_rows_constructor(frame_slice),
    })
}

impl RawMetadataAndRawRows {
///
/// pub fn $deserialize_metadata(
/// &self,
/// ) -> StdResult<DeserializedMetadataAndRawRows<$metadata_l, $raw_rows_kind>, RowsParseError>
    pub fn deserialize_borrowed_metadata(
        &self,
    ) -> StdResult<DeserializedMetadataAndRawRows<'_, RawRowsBorrowed<'_>>, RowsParseError> {
        deserialize_metadata_generic(
            self,
            |frame_slice| RawRowsBorrowed(frame_slice),
            |cached| ResultMetadataHolder::BorrowedOrOwned(Cow::Borrowed(cached)),
            deser_col_specs_borrowed,
        )
    }

    pub fn deserialize_owned_metadata(
        &self,
    ) -> StdResult<DeserializedMetadataAndRawRows<'static, RawRowsOwned>, RowsParseError> {
        deserialize_metadata_generic(
            self,
            |frame_slice| RawRowsOwned(frame_slice.to_bytes()),
            |cached| ResultMetadataHolder::SharedCached(Arc::clone(cached)),
            deser_col_specs_owned,
        )
    }
}
```

Reply:
`generate_deser_col_specs` is also redundant.

Reply:
> In this case I can't deduplicate parts using normal functions because of different ownership/lifetimes. I encourage you to try yourself; I still believe the present state is the best approach here.

After trying myself I removed not only this macro, but all other macros in this file (2).
As the other macros were already on main, I've opened I PR that removes them:
https://github.com/scylladb/scylla-rust-driver/pull/1107

The macro you introduced here is not yet on main, so I've opened a PR to your branch:
https://github.com/wprzytula/scylla-rust-driver/pull/3

Note: I accidentally pushed this commit to your branch (because your branch allows edits by maintainers), but I reverted it quickly. Sorry.